REFRAME.md

```diff
@@ -0,0 +1,290 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Reframe
+
+**Author(s)**:
+- Adin Schmahmann
+- Petar Maymounkov
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+The Reframe protocol is designed for request-response messages that is sufficiently generic and extensible to evolve over time as new needs for it arise. This includes separately defining the transport and message serialization mechanisms from the actual method types and semantics.
+
+The initial use case motivating the protocol's design is to help peers discover various routing hints that enable them to find the content or service they are actually looking for. The extensibility required in this case is that various routing systems could share the protocol so that applications can work on them generically and that recursive routing systems could choose to use the protocol themselves to maximize interoperability.
+
+# Organization of this document
+
+- [Introduction](#introduction)
+- [Spec](#spec)
+- [Intended Upgrade Paths](#intended-upgrade-paths)
+- [Implementations](#implementations)
+
+# Introduction
+
+The Reframe protocol is a request-response based protocol. Upon receiving a request a Reframe server should respond to clients with information they have pertaining to the request. It's possible that a Reframe server may not have all of the information necessary for a client to actually get the data they need, but even partial resolution is acceptable. For example, if a peer is looking to download some block of data a Reframe server may only know about some peers that have the data but not their network addresses. That's ok and the client can always choose alternative mechanisms to discover the missing information. 
+
+# Spec
+
+To build an implementation of the protocol it is required to both define a transport for the messages and the method types supported by the given implementation. Over time both the method and transport types may grow and also be included in this specification.
+
+## Interaction Pattern
+
+Given that a client C wants to request information from some server S:
+
+1. C opens sends a request to S that is a single message
+2. S will respond to C with either a single message or a group of messages
+
+## Transports
+
+### HTTP + DAG-JSON
+
+All messages MUST be encoded as DAG-JSON.
+
+Requests MUST be sent as HTTP POST requests to the endpoint where the message data is sent in the body with chunked encoding.

```

Comment:
does this disallow `HTTP/2` transport upgrades, which subsume chunked encoding with it's own transport framing?

Reply:
@lidel @guseggert I think you all have more context in HTTP land than I do so figure you all might have better answers here

Reply:
It doesn't disallow transport upgrades per se, b/c the req is still handled as HTTP/1.1 on an upgrade, whereas the response is sent back on a half-closed HTTP/2 stream. But it disallows connection reuse b/c you can't make subsequent reqs using HTTP/1.1 chunked encoding on an HTTP/2 stream. It also disallows https with HTTP/2, because in that case HTTP/2 uses ALPN instead of Upgrade headers for negotiation, so the initial req is sent as HTTP/2. But all this goes away if we don't require HTTP/1.1 chunked-encoded requests, which seems reasonable to me.